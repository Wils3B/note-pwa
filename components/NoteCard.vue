<template>
  <v-card
    :class="'lt-' + note.color"
    :to="`/note-detail/${note.id}`"
    class="note-card"
  >
    <v-img v-if="firstImage" :src="firstImage" height="100px"></v-img>
    <v-card-text class="main-text">
      <p v-text="noteText" />
    </v-card-text>
    <v-spacer />
    <v-card-text class="footer-text">
      <span>
        {{ modifiedDate }}
      </span>
      <v-icon v-if="note.voices.length" size="16">mic</v-icon>
    </v-card-text>
  </v-card>
</template>

<script>
import Note from '~/model/note'

export default {
  name: 'NoteCard',
  props: {
    note: {
      type: Object,
      default: () => new Note('Welcome to Note PWA')
    }
  },
  computed: {
    noteText() {
      if (!this.note.id || !this.note.content) return ''
      let str = Note.textContent(this.note)
      str = str[0].toUpperCase() + str.slice(1)
      if (Note.hasImage(this.note) && str.length > 22)
        return `${str.slice(0, 18)} ...`
      else if (!Note.hasImage(this.note) && str.length > 95)
        return `${str.slice(0, 92)} ...`
      return str
    },
    modifiedDate() {
      const date = new Date(this.note.modifiedAt)
      const now = new Date()
      if (
        date.getUTCDay() === now.getUTCDay() &&
        date.getUTCMonth() === now.getUTCMonth() &&
        date.getUTCFullYear() === now.getUTCFullYear()
      ) {
        return date.toTimeString().slice(0, 5)
      }
      return this.note.modifiedAt.slice(0, 10)
    },
    firstImage() {
      if (Note.hasImage(this.note)) return Note.getFirstImage(this.note)
      return undefined
    }
  }
}
</script>

<style lang="scss">
.note-card.v-card {
  height: 100%;
  max-height: 160px;
  overflow: hidden;
  display: flex;
  flex-direction: column;

  &.lt-purple {
    background-color: $lt-purple;
  }
  &.lt-yellow {
    background-color: $lt-yellow;
  }
  &.lt-green {
    background-color: $lt-green;
  }
  &.lt-brown {
    background-color: $lt-brown;
  }
  &.lt-red {
    background-color: $lt-red;
  }
  &.lt-blue {
    background-color: $lt-blue;
  }

  .v-card__text {
    padding: 8px 8px 2px 8px;

    &.main-text > p {
      height: 100%;
      max-height: 100%;
      margin: 0;
      text-overflow: ellipsis;
      overflow: hidden;
      padding-bottom: 0px;
      line-height: 1.4;
    }

    &.footer-text {
      font-size: 12px;
      padding-top: 2px;
      display: flex;
      justify-content: space-between;
      font-weight: 700;
    }
  }
}
</style>
