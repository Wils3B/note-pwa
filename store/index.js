export const state = () => {
  return {
    colors: ['white', 'yellow', 'purple', 'green', 'red', 'blue', 'brown'],
  }
}
