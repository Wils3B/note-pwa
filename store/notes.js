import DB from '~/model/db'
import Note from '~/model/note'

export const state = () => {
  const notes = []
  return {
    notes,
  }
}

export const mutations = {
  saveNote(currentState, note) {
    currentState.notes.push(note)
    DB.addNotes(note)
  },
  deleteNote(currentState, id) {
    const note = currentState.notes.filter((note) => note.id === id)[0]
    const index = currentState.notes.indexOf(note)
    currentState.notes.splice(index, 1)
    DB.deleteNotes(id)
  },
  editNote(currentState, note) {
    const index = currentState.notes.indexOf(
      getters.noteById(currentState)(note.id)
    )
    const currentNote = currentState.notes[index]
    currentNote.color = note.color
    Note.setContent(currentNote, note.content)
    currentState.notes.splice(index, 1)
    currentState.notes.push(currentNote)
    DB.putNotes(note)
  },
  setAllNotes(currentState, notes) {
    currentState.notes = notes
  },
}

export const getters = {
  noteById(currentState) {
    window.notes = currentState.notes
    return (id) => currentState.notes.filter((note) => note.id === id)[0]
  },
}

export const actions = {
  async fetchAllNotes(context) {
    if (!DB.connected) await DB.connect()
    let notes
    if (DB.connected) {
      notes = await DB.getAllNotes()
      context.commit(
        'setAllNotes',
        notes.data.sort((a, b) => (a.modifiedAt < b.modifiedAt ? -1 : 1))
      )
    }
  },
}
