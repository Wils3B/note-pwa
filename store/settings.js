export const state = () => {
  const loadedSettings = loadSettings()
  return {
    selecteds: {
      sortType: loadedSettings.sortType || 'date'
    },
    groups: {
      sortTypes: ['date', 'color']
    }
  }
}

export const mutations = {
  changeSortType(currentState, type) {
    currentState.selecteds.sortType = type
    storeSettings(currentState.selecteds)
  }
}

function storeSettings(settings) {
  localStorage.setItem('settings', JSON.stringify(settings))
}

function loadSettings() {
  const settingsJSON = localStorage.getItem('settings')
  let settingsObj = {}
  try {
    settingsObj = JSON.parse(settingsJSON)
  } catch (e) {}
  return settingsObj || {}
}
