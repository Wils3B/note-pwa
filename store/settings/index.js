import { loadSettings, storeSettings } from './functions'

export const state = () => {
  const loadedSettings = loadSettings()
  return {
    selecteds: {
      sortType: loadedSettings.sortType || 'date',
      defaultColor: loadedSettings.defaultColor || 'white',
      appFont: loadedSettings.appFont || 'Amita',
      autosave: loadedSettings.autosave || '2min',
    },
    groups: {
      sortTypes: ['date', 'color'],
      fonts: ['Kalam', 'Amita', 'Roboto'],
      autosave: ['off', '10s', '20s', '30s', '2min', '5min'],
    },
  }
}

export const mutations = {
  changeSortType(currentState, type) {
    currentState.selecteds.sortType = type
    storeSettings(currentState.selecteds)
  },
  changeDefaultColor(currentState, color) {
    currentState.selecteds.defaultColor = color
    storeSettings(currentState.selecteds)
  },
  changeAppFont(currentState, font) {
    currentState.selecteds.appFont = font
    storeSettings(currentState.selecteds)
  },
  changeAutosave(currentState, autosave) {
    currentState.selecteds.autosave = autosave
    storeSettings(currentState.selecteds)
  },
}

export const getters = {
  autosaveInMillis(currentState) {
    const setting = currentState.selecteds.autosave
    if (/^\d+s$/.test(setting)) {
      return 1000 * Number(setting.substring(0, setting.length - 1))
    } else if (/^\d+min$/.test(setting)) {
      return 1000 * 60 * Number(setting.substring(0, setting.length - 3))
    }
    return 0
  },
}
