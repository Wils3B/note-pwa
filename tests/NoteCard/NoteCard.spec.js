import { mount } from '@vue/test-utils'
import Vuetify from 'vuetify'
import { localInstance } from '../common'
import router from '../router'
import store from '../store'
import notes from '../notes'
import NoteCard from '~/components/NoteCard.vue'

const localVue = localInstance()

describe('Note Card', () => {
  const wrapper = mount(NoteCard, {
    localVue,
    router,
    store,
    vuetify: new Vuetify(),
    propsData: {
      note: notes[0],
    },
  })

  test('it should be a vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy()
  })

  test('Have a correct render', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('it should have good display', () => {
    expect(wrapper.find('.v-image').exists()).toBe(true)
    expect(wrapper.find('.main-text').text()).toBe('Note 1583267092854')
    expect(wrapper.find('.footer-text').text()).toBe('2020-03-03')
  })

  test('it should be a link to note details', () => {
    expect(wrapper.attributes('href')).toBe(`#/note-detail/${notes[0].id}`)
  })

  test('it should have good display', () => {
    expect(wrapper.find('.v-image').exists()).toBe(true)
    expect(wrapper.find('.main-text').text()).toBe('Note 1583267092854')
    expect(wrapper.find('.footer-text').text()).toBe('2020-03-03')
  })

  test('it should have a default display', () => {
    const localWrapper = mount(NoteCard, {
      localVue,
      router,
      store,
    })
    expect(localWrapper.find('.v-image').exists()).toBe(false)
    expect(localWrapper.find('.main-text').text()).toBe('Welcome to Note PWA')
    expect(localWrapper.find('.footer-text').text()).toBe(new Date().toTimeString().slice(0, 5))
  })
})
