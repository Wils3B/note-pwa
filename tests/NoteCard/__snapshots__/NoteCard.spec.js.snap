// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note Card Have a correct render 1`] = `
"<a href=\\"#/note-detail/1583267092854\\" class=\\"note-card v-card v-card--link v-sheet theme--light lt-white\\" tabindex=\\"0\\">
  <div class=\\"v-image v-responsive theme--light\\" style=\\"height: 100px;\\">
    <transition-stub name=\\"fade-transition\\" mode=\\"in-out\\">
      <div class=\\"v-image__image v-image__image--preload v-image__image--cover\\" style=\\"background-position: center center;\\"></div>
    </transition-stub>
    <div class=\\"v-responsive__content\\"></div>
  </div>
  <div class=\\"v-card__text main-text\\">
    <p>Note 1583267092854 </p>
  </div>
  <div class=\\"spacer\\"></div>
  <div class=\\"v-card__text footer-text\\"><span>
      2020-03-03
    </span>
    <!---->
  </div>
</a>"
`;
